### PR TITLE
[RzIL] Make RzILOp statically typed

### PR DESCRIPTION
**Pull request by thestr4ng3r**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This makes IL ops statically typed, such that not all handlers have to explicitly supply their return type at runtime and it is much clearer what sort of op is expected where.
In particular, the old `RzILOp` has been split into `RzILOpEffect` (`'a effect`) and `RzILOpPure` (`'a pure`). Pure effects are then subdivided into bool (`bool`) and bitvector (`'a bitv`) ops, but this cannot be expressed statically in a trivial fashion, as there exist polymorphic ops like `ite : bool -> 'a pure -> 'a pure -> 'a pure`, so here we resort to just the typedefs `RzILOpBool` and `RzILOpBitVector`.
As a side-effect, the implicit casts in `rz_il_evaluate_*` have been removed (they would have caused seroious issues later on in analysis anyway), and the `eq` and `is_zero` ops (`non_zero` too as a helper, but not a real op) has been introduced to preserve the current functionality.

**Test plan**

run the tests

### CODE REVIEW

File: librz/il/vm_layer.c [hunk 1]
```diff
@@ -4,102 +4,107 @@
...
-	rz_il_handler_seq, /* RZIL_OP_SEQ */
-	rz_il_handler_unimplemented, // &rz_il_handler_blk, /* RZIL_OP_BLK */
-	rz_il_handler_unimplemented, // &rz_il_handler_repeat, /* RZIL_OP_REPEAT */
-	rz_il_handler_branch, /* RZIL_OP_BRANCH */
-	rz_il_handler_unimplemented, // &rz_il_handler_invalid, /* RZIL_OP_INVALID */
+void *rz_il_handler_pure_unimplemented(RzILVM *vm, RzILOpPure *op, RzILPureType *type);
+bool rz_il_handler_effect_unimplemented(RzILVM *vm, RzILOpEffect *op);
+
+static RzILOpPureHandler op_handler_pure_table_default[RZIL_OP_PURE_MAX] = {
+	[RZIL_OP_VAR] = rz_il_handler_var,
```
Comment (p371jah) [inline librz/il/vm_layer.c:61]: Nice

File: test/unit/test_il_vm.c [hunk 2]
```diff
@@ -143,28 +143,27 @@ static bool test_rzil_vm_root_evaluation() {
...
-	RzILOp *arg1 = rz_il_op_new_bitv_from_st64(16, 23);
-	RzILOp *arg2 = rz_il_op_new_bitv_from_st64(16, 19);
-	RzILOp *add = rz_il_op_new_add(arg1, arg2);
-	RzILOp *true_val = rz_il_op_new_b1();
-	RzILOp *false_val = rz_il_op_new_b0();
-	RzILOp *ite_root = rz_il_op_new_ite(add, true_val, false_val);
+	RzILOpBitVector *arg1 = rz_il_op_new_bitv_from_st64(16, 23);
+	RzILOpBitVector *arg2 = rz_il_op_new_bitv_from_st64(16, 19);
+	RzILOpBitVector *add = rz_il_op_new_add(arg1, arg2);
+	RzILOpBool *condition = rz_il_op_new_bool_inv(rz_il_op_new_eq(add, rz_il_op_new_bitv_from_ut64(16, 0)));
```
Comment (wargio) [inline test/unit/test_il_vm.c:149]: so now bitv are not automatically converted to bool ?

Comment (thestr4ng3r) [reply test/unit/test_il_vm.c:149]: Correct. Bitvectors and bool are two different things and BAP does not do this either. This particular example is a bit overcomplicated for testing, usually you will just want to use `rz_il_op_new_non_zero()`.

Comment (wargio) [reply test/unit/test_il_vm.c:149]: this is kinda a problem for me, but i do understand the point.

File: librz/il/rzil_opcodes.c [hunk 3]
```diff
@@ -655,128 +676,147 @@ RZ_API void rz_il_op_free(RZ_NULLABLE RzILOp *op) {
...
-		rz_il_op_free_1(set, x);
+		rz_il_op_free_1(pure, set, x);
 		break;
 	case RZIL_OP_LET:
-		rz_il_op_free_1(let, x);
+		rz_il_op_free_1(pure, let, x);
 		break;
 	case RZIL_OP_JMP:
-		rz_il_op_free_1(jmp, dst);
+		rz_il_op_free_1(pure, jmp, dst);
```
Comment (XVilka) [inline librz/il/rzil_opcodes.c:802]: Is this on purpose¿

File: librz/include/rz_il/rzil_vm.h [hunk 4]
```diff
@@ -16,17 +17,26 @@ extern "C" {
...
+	RZ_IL_PURE_TYPE_BOOL, ///< RzILBool
+	RZ_IL_PURE_TYPE_BITV ///< RzBitVector
+} RzILPureType;
 
 typedef struct rz_il_vm_t RzILVM;
-typedef void *(*RzILOpHandler)(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
-typedef void (*RzILVmHook)(RzILVM *vm, RzILOp *op);
+
+/**
+ * \brief Evaluation callack for a single pure opcode
```
Comment (XVilka) [inline librz/include/rz_il/rzil_vm.h]: A typo

File: librz/include/rz_il/rzil_vm.h [hunk 5]
```diff
@@ -16,17 +17,26 @@ extern "C" {
...
+ * \param type when returning a non-null value, this must be set to the respective type.
+ * \return The evaluated value of the type indicated by *type, or NULL if an error occured and the execution should be aborted
+ */
+typedef void *(*RzILOpPureHandler)(RzILVM *vm, RzILOpPure *op, RZ_NONNULL RZ_OUT RzILPureType *type);
+
+/**
+ * \brief Evaluation (execution) callack for a single effect opcode
```
Comment (XVilka) [inline librz/include/rz_il/rzil_vm.h]: A typo

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/il/rzil_opcodes.c (hunk 3 above)
Comment (thestr4ng3r) [reply librz/il/rzil_opcodes.c:802]: I'm not sure what you are referring to. The `rz_il_op_free_1` macro receives a new argument to select the right functions for the type.

Comment (thestr4ng3r) [reply librz/il/rzil_opcodes.c:802]: `jmp->dst` is pure.

**Pull request merged**